src/Orleans.Core/Runtime/ExecutorService.cs

```diff
@@ -1,15 +1,54 @@
-﻿using System.Threading.Tasks;
-using Orleans.Runtime.Scheduler;
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using Orleans.Messaging;
 
 namespace Orleans.Runtime
 {
-    internal class ExecutorService : ITaskScheduler
+    internal interface IExecutor
     {
-        private readonly TaskScheduler taskScheduler = new ThreadPerTaskScheduler(task => (task as AsynchAgentTask)?.Name);
+        void QueueWorkItem(WaitCallback callback, object state = null);
 
-        public void RunTask(Task task)
+        int WorkQueueCount { get; }
+    }
+
+    internal class ExecutorService
+    {
+        public IExecutor GetExecutor(GetExecutorRequest request)
+        {
+            if (typeof(GatewayConnection).IsAssignableFrom(request.StageType))
+            {
+                // After stop GatewayConnection needs to reroute not yet sent messages to another gateway 
+                return ConstructQueuedExecutor()(true);
+            }
+
+            if (typeof(SingleTaskAsynchAgent).IsAssignableFrom(request.StageType))
+            {
+                return new ThreadPerTaskExecutor(request.StageName);
+            }
+
+            return ConstructQueuedExecutor()(false);
+
+            Func<bool, QueuedExecutor> ConstructQueuedExecutor()
```

Comment:
This method might be clearer if it just returns the executor immediately, like this:
```C#
QueuedExecutor ConstructQueuedExecutor(bool drainAfterCancel)
{
  return new QueuedExecutor(request.StageName, request.CancellationTokenSource, drainAfterCancel);
}
```

And then call it specifying the argument name, `drainAfterCancel`.

Alternatively, DrainAfterCancel might be a good candidate to move into the request, since it seems like it's important for the caller to know what the behavior is.

Reply:
Indeed, the `drainAfterCancel` usage here is just leaking knowledge about `GatewayConnection` implementation.